evenless.1
\" Copyright (C) 2011 George Silvis, III <george.iii.silvis@gmail.com>

\" This file is part of evenless.
\"
\" evenless is free software: you can redistribute it and/or modify it under the
\" terms of the GNU General Public License as published by the Free Software
\" Foundation, either version 3 of the License, or (at your option) any later
\" version.
\"
\" evenless is distributed in the hope that it will be useful, but WITHOUT ANY
\" WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
\" A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
\"
\" You should have received a copy of the GNU General Public License along with
\" evenless.  If not, see <http://www.gnu.org/licenses/>.


.TH EVENLESS 1 evenless\-0.0.5
.SH NAME
evenless \- front-end to the notmuch mail indexer
.SH SYNOPSIS
.B evenless
[\fIOPTIONS\fR]
.SH DESCRIPTION
.BR evenless
is a minimalistic frontend to the
.BR notmuch(1)
mail-indexing system. It lets you search for, tag, and untag threads. For all
actions involving individual messages (composing and reading), it invokes
.BR mutt(1)
with the appropriate arguments.
.SH OPTIONS
.TP
.B [\-c/\-\-config] /path/to/config
Uses /path/to/config as the configuration file rather than the default
~/.evenless/evenless.conf
.TP
.B [\-d/\-\-database] /path/to/notmuch/database
Uses /path/to/notmuch/database as the database file, rather than reading a
config file to find it.
.TP
.B [\-s/\-\-search] query
Uses query as the search term rather than the default "tag:inbox". See
.BR notmuch(1)
for more details.
.TP
.B \-\-version
Prints the version and copyright, and then exits.
.SH USAGE
.TP
.B c/m
Compose a new message
.TP
.B q/x
Exit immediately.
.TP
.B u/f
Fetch new messages. 
.TP
.B t
Select a thread, then add a tag to it
.TP
.B T
Select a thread, then remove a tag from it
.TP
.B a
Select a thread, then remove the "inbox" tag from it
.TP
.B r
Select a thread, then remove the "unread" tag from it
.TP
.B A
Select a thread, then remove both the "inbox" and "unread" tags from it
.TP
.B v/Enter
Select a thread, then view the entire thread in
.BR mutt(1)
.SH FILES
.TP
.B ~/.evenless/evenless.conf
This is the default configuration file for evenless, though it can be
overrided with the [-c/--config] option. As of right now this file has very
simple syntax: Lines beginning with '#' are comments. There must be exactly
one non-commented line, which specifies the full path to a notmuch database.
.SH SEE ALSO
.BR notmuch(1)
.BR mutt(1)
.BR offlineimap(1)
.BR msmtp(1)
.SH LIMITATIONS
.BR evenless
does not have the capability of changing the tags of an individual message.
.BR evenless
also does not currently handle or even think about syncing notmuch tags to
Maildir flags. This will be fixed later.
.br evenless
currently acts on the actively loaded database.  Instead, a buffer should be
used, and updates flushed to the database later.
.SH AUTHOR
Copyright (C) 2011 George Silvis, III.
.PP
Permission is granted to copy, distribute and/or modify this document under
the terms of the GNU General Public License, Version 3 or any later version
published by the Free Software Foundation.


CHANGELOG
evenless-0.0.5
        Bug fixes:
          Viewing a thread marks it as read.
        Switched from cmake to autotools. (Got fed up with cmake.)
          install and uninstall targets should both work, among many more.
        Added 'u/f' command: run offlineimap and notmuch new.
        ncurses support in a usable state.

evenless-0.0.4-rc1
        Complete re-structuring of code, allowing many planned features.
        I love realloc.

evenless-0.0.3
	Distribution changes:
	  Migrated to cmake (build systems, woo!)
          Added copyright notices to evenless.1 and CMakeLists.txt (Thank you, Liam!)
	Bug fixes:
	  Compiles w/out warnings
          Realized that it's not 2010 anymore, and that these files should probably say copyright 2011

evenless-0.0.2
        Distribution changes:
          Added a header file evenless.h
          Added a man page evenless.1
          Added a disclaimer to the --version output
        New features:          
          Added '-d/--database' command-line option
          Added 'A' key to remove both 'inbox' and 'unread' tags
          Threads with unread messages are printed in BOLD
        Bug fixes:
          Fixed bug where selecting threads would sometimes fail [not fully understood yet]
          String input now strips trailing newline, fixing a bug where tagging would seem to fail
        New limitations realized:
          Syncing tags to maildir flags is not handled in any way currently.

evenless-0.0.1
        First release.


src/evenless.c
/* Copyright (C) 2011 George Silvis, III <george.iii.silvis@gmail.com> */

/*
This file is part of evenless.

evenless is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

evenless is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
evenless.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <notmuch.h>
#include <ncurses.h>

#include "consts.h"
#include "types.h"

#include "config.h"

#include "get.h"
#include "utils.h"
#include "actions.h"


#include "evenless.h"

/* Throughout this file, if not mentioned otherwise, any function that returns an int
 * returns 0 if successful, and 1 if unsuccessful
 */


void
print_thread (Client *c,
              int i) {

  if (i >= c->num_entries) {
    clrtoeol();
    printw("\n");
    return;
  }
  
  notmuch_thread_t *thread = c->entries[i];
    
  notmuch_tags_t *tags;
  int unread = FALSE;
  for (tags = notmuch_thread_get_tags(thread);
       notmuch_tags_valid(tags);
       notmuch_tags_move_to_next(tags)) {
    const char *tag = notmuch_tags_get(tags);
    if (!strcmp(tag, "unread")) {
      unread = TRUE;
      break;
    }
  }
  notmuch_tags_destroy(tags);
  
  int total_messages = notmuch_thread_get_total_messages(thread);
  const char *authors = notmuch_thread_get_authors(thread);
  const char *subject = notmuch_thread_get_subject(thread);
  /* printf("\e[%im%i: %i messages. %s.  %s\e[m\n", style, i, total_messages, subject, authors); */
  if (unread) {
    attron(A_BOLD);
  }
  if (i==c->selected) {
    attron(A_REVERSE);
  }
  printw("%i: %i messages. %s. %s \n", i, total_messages, subject, authors);
  clrtoeol();
  attroff(A_BOLD);
  attroff(A_REVERSE);
  return;

}



void
print_ttyui (Client* c) {

  /*  get_entries(c); */
  int i;

  int y = getmaxy(stdscr);

  move(0, 0);

  if (c->top_displayed > c->selected) {
    c->top_displayed -= y/3;
  }

  if (c->top_displayed + y - 2 < c->selected) {
    c->top_displayed += y/3;
  }


  /* fix top_displayed */
  if (c->top_displayed < 0) {
    c->top_displayed = 0;
  } else if (c->num_entries <= y+1) {
    c->top_displayed = 0;
  } else if (c->top_displayed + y >= c->num_entries) {
    c->top_displayed = c->num_entries - y - 1;
  }



  for (i = c->top_displayed;
       i < y+(c->top_displayed)-1 /* && i < c->num_entries */;
       i++) {
    print_thread(c, i);
  }
  refresh();

}


void
run_ttyui (Client *c) {

  char command;

  get_entries(c);

  while (1) {
    print_ttyui(c);
    command = getch();


    switch (command) {
      /* GLOBAL COMMANDS */
    case 'c': case 'm': /* _c_ompose new _m_ail */
      action_compose(c);
      break;
    case 'u': case 'f':
      action_fetch(c);
      get_entries(c);
      break;
    case 'q': case 'x': /* e_x_it */
      return;
      /* THREAD-SPECIFIC COMMANDS */
    case 't': /* select and _t_ag a thread */
      action_tag_thread(c);
      get_entries(c);
      break;
    case 'T': /* select and un_T_ag a thread */
      action_untag_thread(c);
      get_entries(c);
      break;
    case 'a': /* select and archive a thread */
      action_archive(c);
      get_entries(c);
      break;
    case 'r': /* mark a thread as read */
      action_read(c);
      break;
    case 'A': /* marh a thread as read and archive it */
      action_read_archive(c);
      get_entries(c);
      break;
    case 'v': case '\n': /* select and _v_iew a thread */
      action_view(c);
      break;
    case 'j':
      action_move(c, 1);
      break;
    case 'k':
      action_move(c, -1);
      break;
    default:
      output_message("Unknown command.");
    }


  }




}







void
initialize_client (Client *c) {
  c->config_path = NULL;
  c->database_path = NULL;
  c->search_string = "tag:inbox";
  c->database = NULL;
  c->query = NULL;
  c->threads = NULL;
  c->num_entries = 4;
  c->entries = malloc(4 * sizeof(void*));
  c->top_displayed = 1;
  c->selected = 0;
}




void
parse_args (Client *c,
            int argc,
            char **argv) {

  int i;
  for (i = 1; i < argc; i++) {
    if (!strcmp(argv[i], "--version")) {
      printf("%s\n", PACKAGE_STRING);
      printf("Copyright (C) 2011 George Silvis, III\n");
      printf("This is free software; see the source for copying conditions.  There is NO\n");
      printf("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");

      cleanup_client(c);
      free(c);

      exit(0);
    } else if (!strcmp(argv[i], "-s") || !strcmp(argv[i], "--search")) {
      if (i+1==argc) {
        printf("Bad arguments. %s expects a search string (in quotes if necessary).\n", argv[i]);
        exit(1);
      }
      c->search_string = argv[++i];
    } else if (!strcmp(argv[i], "-c") || !strcmp(argv[i], "--config")) {
      if (i+1==argc) {
        printf("Bad arguments. %s expects a path to a file.\n", argv[i]);
        exit(1);
      }
      c->config_path = argv[++i];
    } else if (!strcmp(argv[i], "-d") || !strcmp(argv[i], "--database")) {
      if (i+1==argc) {
        printf("Bad arguments. %s expects a path to a notmuch database.\n", argv[i]);
        exit(1);
      }
      c->database_path = argv[++i];
    } else {
      printf("Bad arguments. (unrecognized argument)\n");
      exit(1);
    }
  }
}

int
cleanup_client (Client *c) {
  if (c->entries) {
    free(c->entries);
  }
  if (c->threads) {
    notmuch_threads_destroy(c->threads);
  }
  if (c->query) {
    notmuch_query_destroy(c->query);
  }
  if (c->database) {
    notmuch_database_close(c->database);
  }
  c->entries = NULL;
  c->num_entries = 0;
  c->threads = NULL;
  c->query = NULL;
  c->database = NULL;

  return 0;
}


int
main (int argc,
      char **argv) {

  Client *c = malloc(sizeof(Client));
  initialize_client(c);

  parse_args(c, argc, argv);

  initscr();
  raw();
  keypad(stdscr, TRUE);
  noecho();
  curs_set(0);

  run_ttyui(c);

  endwin();

  cleanup_client(c);
  free(c);

  return 0;
}
